### PR TITLE
Create very_basic_agent.py

### PR DESCRIPTION
**Pull request by leagonnet33**
implémentation d'un agent extrêmement basique pour me familiariser avec l'environnement 
La logique suivie par cet agent est la suivante:
- si on n'a pas assez d'argent pour acheter, on vend ou on ne fait rien avec proba 0.5
- si on a gagné de l'argent à l'itération précédente (= reward > 1), on achète (ou on ne fait rien avec proba epsilon)
- si on a perdu de l'argent à l'itération précédente (= reward < 1), on vend (ou on ne fait rien avec proba epsilon)

### CODE REVIEW

**Comment by leagonnet33** [comment]
- Création de la classe de l'agent epsilon greedy et d'une fonction permettant de run cet agent. 

- J'ai également ajouté une ligne dans le fichier portfolio car la quantité de cash dans le portefeuille ne se mettait pas à jour quand on faisait un achat

- J'ai géré directement dans la fonction agent_step() les cas où: pas assez de cash pour acheter, ou pas de crypto à vendre. J'ai l'impression qu'il y a des lignes dans portfolio.py qui sont censées gérer ces cas, mais sans ajouter la disjonction de cas dans agent_step() je me retrouvais avec des runs où l'action prise était systématiquement de vendre et on était juste toujours à 0 coin du coup

File: portfolio.py [hunk 1]
```diff
@@ -100,6 +100,8 @@ def __buy(self, current_price):
 
         self.portfolio_coin_ += coin_to_buy
         self.cash_used_ += coin_to_buy * buy_price
+        self.portfolio_cash_ -= coin_to_buy * buy_price
+
```
Comment (bvelitchkine) [inline portfolio.py:104]: Je suis d'accord avec les changements effectués, mais il faut initialiser ```self.portfolio_cash_``` différemment, car pour le moment, c'est initialisé à 0, et nous ce qu'on voudrait, c'est que ce soit initialisé à par exemple 10k€, et que ça baisse à mesure qu'on achète des cryptos, que ça augmente à mesure qu'on en vend.

File: very_basic_agent.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+from environments import *
+from portfolio import *
+from loader import TradingDataLoader
+from random import *
+
+def run_very_basic_agent(epsilon,cash_available = 1000,num_coins_per_order=1, recent_k=0):
+    METRICS = ["open", "high", "low", "close", "volume ETH", "volume USDT", "tradecount"]
```
Comment (bvelitchkine) [inline very_basic_agent.py:7]: On utilise en général cette notation pour des variables globales, définies hors des fonctions, comme dans le fichier `portfolio.py` par exemple.

File: very_basic_agent.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            else: action = 2
+        else:
+            if current_reward >=1:
+                if n > epsilon:
+                    action = 1
+                else: action = 0
+            else: 
+                if n > epsilon:
+                    action = 2
+                else: action = 0
```
Comment (bvelitchkine) [inline very_basic_agent.py:33]: Toute cette partie "prise de décision" devrait aller dans la classe *DumbAgent* que je mentionne dans mes remarques générales.

File: e-greedy_agent.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+def run_epsilon_greedy_agent(initial_cash,epsilon):
+    agent = EpsilonGreedyAgent
+    env_info = {}
+    all_averages = []
+    METRICS = ["open", "high", "low", "close", "volume ETH", "volume USDT", "tradecount"]
+    data = TradingDataLoader().data()
+    env = TradingBotEnv(data,metrics = METRICS)
+    portfolio = Portfolio(portfolio_cash = initial_cash)
+    is_done = False 
+    agent = EpsilonGreedyAgent()
```
Comment (bvelitchkine) [inline e-greedy_agent.py:104]: - **METRICS** hors de la fonction car plutôt une variable globale.
- Instanciation des différents objets dans un fichier `main.py` ou `app.py` cf. remarques générales de cette review.

File: e-greedy_agent.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+            
+        elif cash_amount < current_price and int(coins_amount) > 0: # si on ne peut pas acheter, on vend ou on ne fait rien
+            if test < 0.5: # avec probabilité 1/2
+                current_action = 0
+            else: current_action = 2
+                
+        elif cash_amount >= current_price and int(coins_amount) == 0: # si on ne peut pas vendre, on achète ou on ne fait rien
+            if test < 0.5: # avec probabilité 1/2
+                current_action = 0
+            else: current_action = 1
```
Comment (bvelitchkine) [inline e-greedy_agent.py:75]: Effectivement, cette partie disjonction de cas que tu mentionnes dans tes commentaires est déjà prévues dans la méthode *apply_action* de *Portfolio*. Alors il y a peut-être une erreur dans le code, mais je pense qu'il doit se régler du côté de `portfolio.py` plutôt qu'ici si possible.

**Review by bvelitchkine — CHANGES_REQUESTED** [review]
Je t'ai fait des remarques un peu spécifiques à différents endroits du code, et ici, je liste mes remarques générales :)

Avant tout, je pense qu'il faut revoir l'architecture.

1. Tu as choisi de faire un fichier `e-greedy_agent.py` et un autre `very_basic_agent.py`. Je pense qu'il faut un unique fichier `agents.py` dans lequel sont regroupés tous nos agents, sous forme de classes : le *BaseAgent* que tu as dans `e-greedy_agent.py` tout d'abord, ensuite une classe *DumbAgent* qui reprend ce que tu as fait dans `very_basic_agent.py` et enfin *EpsilonGreedyAgent*.
2. Tout ce qui concerne le run, donc tes fonctions *run_epsilon_greedy_agent* et *run_very_basic_agent* doivent toutes être dans un autre fichier fichier python intitulé `runners.py` ou `trainers.py`. Il y a d'ailleurs sans doute moyen de faire de ces deux fonctions une unique fonction avec les bons paramètres.
3. Il faudrait enfin un dernier fichier `main.py` ou `app.py` dans lequel on fait tourner nos différents agents dans leur environnement. C'est dans ce fichier qu'on instancie les différents objets et qu'on les fait passer dans leurs "trainers".

**Comment by leagonnet33** [comment]
Outdated -> je ferme
